### ISSUE TITLE
Add support for reading password fields from secret manager

### ISSUE THREAD

**Issue by danawillow** [hashicorp/terraform-provider-google#7624]
### Affected Resource(s)

* google_*

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment. If the issue is assigned to the "modular-magician" user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If the issue is assigned to a user, that user is claiming responsibility for the issue. If the issue is assigned to "hashibot", a community member has claimed the issue already.

### Description

https://github.com/hashicorp/terraform-provider-google-beta/pull/2252 is an initial implementation of this for one resource. Filing this issue so that the team can think about what a good solution is that works across resources so we don't end up with a bunch of unrelated oneoffs.

**Comment by upodroid** [comment]
I think we should close this issue, end users can leverage CSEK to encrypt the state if they need to. I showed an example in that issue.

**Comment by rileykarson** [comment]
Labelling this as `question` too since we need to come to a decision on whether we want to take this solution or consider an alternative (for example potentially, as @upodroid points out, the "do nothing" (and wait for Core) option).

**Comment by wyardley** [comment]
I think it would also be useful to be able to put data marked sensitive _into_ secrets manager secrets, for example, the generated auth string returned by Redis https://github.com/hashicorp/terraform-provider-google/issues/7601

If they could be put there directly without also tracking them in the state, that would be a bonus.

> I think we should close this issue, end users can leverage CSEK to encrypt the state if they need to. I showed an example in that issue.

To me, the storage at rest is not really my concern, but more that (as best I know / understand), even with a customer supplied key for the bucket itself, anything that can plan the state will be able to also display the secret in cleartext. Especially for folks using tools like Atlantis I'd say there are some risks here. Yes, the user would also need to have secret manager permissions to encrypt to be able to plan / apply (but I don't think they'd necessarily need decrypt, right?)

For sure, I'd say there'd be some benefit to being able to do the above -- that is, take a secret generated by a Terraform resource (say, a service account credential, or a Redis auth string), and be able to _directly_ store it in GCP secret manager or encrypt it with KMS before tracking it in state.

**Comment by rileykarson** [comment]
> anything that can plan the state will be able to also display the secret in cleartext

Values marked sensitive should be omitted from plans, but will appear as cleartext in provider debug logs. Since those are configured by an environment variables / in config they're readily available to any interactive user but may not be available to automation users.

**Comment by wyardley** [comment]
> Values marked sensitive should be omitted from plans, but will appear as cleartext in provider debug logs

They're cleartext in the raw state files as well too, though, right?

**Comment by rileykarson** [comment]
Yep! Those are harder to access when they're in remote backends, but not much harder.

**Comment by mvanholsteijn** [comment]
Hashicorp Sentinel will handout clear-text secrets as well when you let it create mocks for you.

**Comment by rileykarson** [comment]
Also note https://github.com/hashicorp/terraform-provider-google/issues/927 which captures extensive past discussion about this

**Comment by sscheible** [comment]
Storing secrets unnecessarily is never a good idea, as it may require a lot of effort to protect a secret that does not need to be there at all. When source-of-truth is Secret Manager, Terraform has no need to store a copy of the secret in its own state. Storing some metadata on it is valid however. 

I do understand from [the discussion in 4514](https://github.com/hashicorp/terraform-provider-google/issues/4514#issuecomment-576667692) that Terraform **does** have a need to tell when the secret has changed, so it can mutate the resource using the secret that has changed. I would suggest storing metadata of the secret version in tfstate, to detect when the secret version has changed. [API is offering](https://cloud.google.com/secret-manager/docs/reference/rest/v1beta1/projects.secrets.versions/get) the version behind "latest" and the createTime, which should be good enough to detect any change to the secret's value:
```
GET https://secretmanager.googleapis.com/v1beta1/projects/<yourProject>/secrets/<yourSecret>/versions/latest

{
  "name": "projects/<yourProject>/secrets/<yourSecret>/versions/<numberOfLatestVersion>",
  "createTime": "<secretCreateTimeStamp>",
  "state": "ENABLED"
}
```

Regarding [#927](https://github.com/hashicorp/terraform-provider-google/issues/927#issuecomment-404150509), there are CI/CD use cases where terraform plan can be done locally to achieve "fail fast", but apply is done on a CI/CD server with extra permissions. Planners can view metadata on the secret version, but only the applying CI/CD server can actually read the secret. This protects the secret from planners who do need tfstate access, but have no reasonable need to know the secret.

**Comment by rileykarson** [comment]
HashiCorp has effectively added support for this flow as write-only fields + ephemeral resources. Individual bugs can be filed for given secret source(s) and/or WO fields that are needed.
